### DISCUSSION TITLE
Tips on async tracing

### DISCUSSION THREAD

**Discussion by jeremy-rifkin** [nodejs/help#4619] [Q&A]
I’m trying to debug an issue with a long-running application doing a lot of `async` processing. The issue specifically is that a certain async function is taking much longer than I expect to complete but only rarely. What I would like is a way to record a trace of what async code is running and when. For some code like
```js
async function() {
    foo();
    await bar();
    baz();
}
```
I’d basically like to see when the part before bar() is executed, when bar is called, and when the part after is executed.

Is there any way to do this without manually instrumenting the code?

I have explored some ideas using async_hooks but it seems I can’t quite get the information I want.

**Comment by zebbern** [comment]
Hi! You can try using Node.js’s async_hooks module to trace your async functions without manually adding code everywhere.

quick setup:

```
const async_hooks = require('async_hooks');
const fs = require('fs');

const log = (msg) => fs.writeSync(1, `${msg}\n`);

const hooks = {
  init(id, type) { log(`Start: ${type} (${id})`); },
  before(id) { log(`Before: ${id}`); },
  after(id) { log(`After: ${id}`); },
  destroy(id) { log(`Destroy: ${id}`); }
};

async_hooks.createHook(hooks).enable();
```

This will log each async operation’s start, before, after, and completion events. It’s a minimal setup that should help you see what’s running and when.

Hope this helps!
